### DISCUSSION TITLE
PulseAudio on FreeBSD

### DISCUSSION THREAD

**Discussion by grahamperrin** [poetaman/arttime#26] [General]
<https://github.com/poetaman/arttime/wiki#user-content-unixes> understandably suggests `paplay` as a dependency. 

Where PulseAudio is _required_, it's also commonplace to prevent or avoid use. Typically, configure the client to **not** automatically spawn a daemon when needed: 

`autospawn=no`

### Hint

In this type of environment, before then after temporary use of things such as `paplay`:

1. `pulseaudio --start`
2. `pulseaudio --kill`

### Notes

* <https://www.freedesktop.org/wiki/Software/PulseAudio/>
* <https://www.freshports.org/audio/pulseaudio/#requiredby> in FreshPorts
* [pacat(1)](https://www.freebsd.org/cgi/man.cgi?query=pacat&sektion=1&manpath=FreeBSD-Ports)
* [pulseaudio(1)](https://www.freebsd.org/cgi/man.cgi?query=pulseaudio&sektion=1&manpath=FreeBSD-Ports)
* [pulse-client.conf(5)](https://www.freebsd.org/cgi/man.cgi?query=pulse-client.conf&sektion=5&manpath=FreeBSD-Ports) – [DIRECTIVES](https://www.freebsd.org/cgi/man.cgi?query=pulse-client.conf&sektion=5&manpath=FreeBSD-Ports#DIRECTIVES) include `autospawn`

FreeBSD bugs: 

* [194727 – uaudio device gets disconnected, and hangs usb until everything using /dev/mixer* is closed](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=194727)
* [262713 – audio/pulseaudio: update to 15.0 or greater](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=262713)

```text
% file /usr/local/bin/paplay
/usr/local/bin/paplay: symbolic link to pacat
% pkg provides /usr/local/bin/paplay
Name    : pulseaudio-14.2_4
Comment : Sound server for UNIX
Repo    : FreeBSD
Filename: usr/local/bin/paplay
% pkg rquery %o pulseaudio
audio/pulseaudio
% 
```

**Comment by poetaman** [comment]
@grahamperrin  Thanks! Is there another more native way to play notification sounds in DE on BSD Unixes then? All of my Linux machines (Arch, Debian, Kali) with Desktop Environments it comes enabled by default or is enabled by default once I install a package for pulseaudio. I never had to or cared about turning it on and off. After all its a daemon, and when its not being invoked its not consuming any CPU time... I see the FreeBSD bug that could restrict a user from using pulseaudio, in that case another method of playing sound could be added as a dependency and used by arttime till pulseaudio issues are fixed in FreeBSD.

Arttime shouldn't silently turn the daemon on and off for the user. Consider this: A user starts arttime, arttime determines that the pulseaudo daemon is not running. Arttime decides to start the daemon for the user with the intent to close it while it exits (given it was the one who started it). Meanwhile the user turns on their music player that also uses pulseaudio daemon. Then in the sequence of events they quit arttime before they want to stop playing music. Arttime will kill the music being played by the user, and annoying the user...

**Reply by grahamperrin** [reply]
> … Arttime shouldn't silently turn the daemon on and off …

True. 

----

> … Is there another more native way to play notification sounds in DE on BSD Unixes then? …

Spun off to https://github.com/poetaman/arttime/discussions/29 – I can't predict how long it might become …

**Reply by grahamperrin** [reply]
For clarity: I mean, I agree that (on FreeBSD) arttime should not take the `pulseaudio --start` … `pulseaudio --kill` approach.

**Comment by poetaman** [comment]
@grahamperrin On Linux VM this is what I see: the reason``paplay`` is not affected by ``pulseaudio --start`` and ``pulseaudio --kill`` is that it seems to first start a ``pulseaudio`` daemon if it was not already running. Here's a change I made: 
* If pulseaudio is installed and running, then use pulse audio.
* Else if pulseaudio is either not installed or not running, use ogg123.

This fixed multiple problems that could have arisen from prior logic for both Linux and FreeBSD:
* If pulseaudio is not already running, we don't accidentally start the daemon. It's unfortunate that ``paplay`` implicitly starts pulseaudio daemon and does not document it on its man page.‡
* If pulseaudio daemon is running, prefer paplay over ogg123, as it seems to produce clearer sound on some systems.

‡ Note: This fixes any problem arttime could have caused on Linux/BSD systems that had pulseaudio installed and disabled, because  user was using a competing daemon.

Fix went into main https://github.com/poetaman/arttime/commit/0c2a0cbb8f724e25b229f826ba855d9efeec95a0, pull and give it a try. I'll send an email to pulse audio developers to mention on their man page for paplay that paplay implicitly starts pulseaudio daemon, as if user did ``pulseaudio --start`` first.

**Reply by grahamperrin** [reply]
> … Fix went into main [0c2a0cb](https://github.com/poetaman/arttime/commit/0c2a0cbb8f724e25b229f826ba855d9efeec95a0), pull and give it a try. …

1. `pkg delete -f vorbis-tools`
2. `nano /usr/local/etc/pulse/client.conf`
3. `autospawn = no`
4. save
5. `reboot -r`
6. `bin/arttime`
7. tested OK
8. `nano /usr/local/etc/pulse/client.conf`
9. `; autospawn = no`
10. save
11. `pkg check -d`
12. automated reinstallation of `vorbis-tools`
13. `reboot -r`

**Reply by poetaman** [reply]
@grahamperrin Awesome! If I got it right, it works as desired under all your setup conditions (I did similar testing on Linux side).

Am new to FreeBSD so a question:``pkg check -d`` does sanity check and also reinstalls ``vorbis-tools``? (perhaps because it's a dependency of some other package you have)?

**Reply by grahamperrin** [reply]
In my case: 

```text
% pkg_tree -r vorbis-tools
vorbis-tools-1.4.2_1,3
 \__ cdrtools-2022.09.18
% 
```

[sysutils/pkg_tree](https://www.freshports.org/sysutils/pkg_tree/)

[pkg-delete(8)](https://www.freebsd.org/cgi/man.cgi?query=pkg-delete&sektion=8&manpath=FreeBSD-Ports)

[pkg-check(8)](https://www.freebsd.org/cgi/man.cgi?query=pkg-check&sektion=8&manpath=FreeBSD-Ports)

**Comment by poetaman** [comment]
Now [PipeWire](https://pipewire.org) (the supposed drop-in replacement for PulseAudio) is also supported since https://github.com/poetaman/arttime/commit/b4ff7e857889466070ffce05ab85c6b7a593722e. Not sure how it performs on FreeBSD. It is the default on Fedora, and works well there.

**Comment by grahamperrin** [comment]
Related (but not to arrtime): 

* https://gist.github.com/grahamperrin/f741050b2dac26ed79e8a3f1748302e5

– via [FreeBSD sleep/wake (suspend/resume): extending rc.suspend to allow detachment of storage devices on USB, and to avoid audio-related issues](https://old.reddit.com/r/freebsd/comments/xw1a7e/freebsd_sleepwake_suspendresume_extending/)
